File: python/arp-sdk/arp_sdk/obligations.py
"""
Egress obligation pipeline — mirrors
`@kybernesis/arp-sdk/src/obligations.ts`.

Supported obligation types (same list as the TS SDK for wire parity):
  - redact_fields
  - redact_fields_except
  - redact_regex
  - summarize_only
  - aggregate_only
  - insert_watermark
  - no_downstream_share

Non-payload obligations (rate_limit, require_fresh_consent, charge_usd, ...)
pass through — they're enforced elsewhere in the stack.
"""

from __future__ import annotations

import copy
import json
import re
from typing import Any, Callable, Iterable, Optional

from .types import Obligation


def apply_obligations(
    payload: Any,
    obligations: Iterable[Obligation],
    on_unknown: Optional[Callable[[str, dict[str, Any]], None]] = None,
) -> Any:
    """Apply the supplied obligations to `payload`. Returns a new value."""
    current = copy.deepcopy(payload)
    for o in obligations:
        current = _apply_one(current, o, on_unknown)
    return current


def _apply_one(
    payload: Any,
    obligation: Obligation,
    on_unknown: Optional[Callable[[str, dict[str, Any]], None]],
) -> Any:
    t = obligation.type
    params = obligation.params or {}
    if t == "redact_fields":
        return _redact_fields(payload, _string_list(params.get("fields")))
    if t == "redact_fields_except":
        return _redact_fields_except(
            payload, _string_list(params.get("fields"))
        )
    if t == "redact_regex":
        pattern = params.get("pattern")
        if not isinstance(pattern, str):
            return payload
        replacement = params.get("replacement")
        return _redact_regex(
            payload, pattern, replacement if isinstance(replacement, str) else "[redacted]"
        )
    if t == "summarize_only":
        max_words = params.get("max_words")
        if not isinstance(max_words, int):
            max_words = 50
        return _summarize_only(payload, max_words)
    if t == "aggregate_only":
        return _aggregate_only(payload)
    if t == "insert_watermark":
        return _add_marker(payload, "_watermark", params)
    if t == "no_downstream_share":
        return _add_marker(payload, "_no_downstream_share", True)
    if t in {
        "rate_limit",
        "require_fresh_consent",
        "require_vc",
        "log_audit_level",
        "delete_after",
        "notify_principal",
        "charge_usd",
    }:
        return payload
    if on_unknown:
        on_unknown(t, params)
    return payload


def _string_list(value: Any) -> list[str]:
    if not isinstance(value, list):
        return []
    return [v for v in value if isinstance(v, str)]


def _redact_fields(payload: Any, fields: list[str]) -> Any:
    if not fields:
        return payload
    out = copy.deepcopy(payload)
    for path in fields:
        _delete_path(out, path.split("."))
    return out


def _redact_fields_except(payload: Any, fields: list[str]) -> Any:
    if not isinstance(payload, dict):
        return payload
    if not fields:
        return {}
    allow = set(fields)
    return {k: v for k, v in payload.items() if k in allow}


def _redact_regex(payload: Any, pattern: str, replacement: str) -> Any:
    try:
        regex = re.compile(pattern)
    except re.error:
        return payload

    def walk(v: Any) -> Any:
        if isinstance(v, str):
            return regex.sub(replacement, v)
        if isinstance(v, list):
            return [walk(i) for i in v]
        if isinstance(v, dict):
            return {k: walk(val) for k, val in v.items()}
        return v

    return walk(payload)


def _summarize_only(payload: Any, max_words: int) -> Any:
    text = payload if isinstance(payload, str) else json.dumps(payload, default=str)
    words = [w for w in text.split() if w]
    summary = " ".join(words[:max_words])
    return {"summary": summary, "_obligation": "summarize_only", "max_words": max_words}


def _aggregate_only(payload: Any) -> Any:
    if isinstance(payload, list):
        return {
            "count": len(payload),
            "aggregate": "count",
            "_obligation": "aggregate_only",
        }
    if isinstance(payload, dict):
        for k, v in payload.items():
            if isinstance(v, list):
                return {
                    k: {"count": len(v), "aggregate": "count"},
                    "_obligation": "aggregate_only",
                }
    return {"aggregate": "count", "count": 1, "_obligation": "aggregate_only"}


def _add_marker(payload: Any, key: str, value: Any) -> Any:
    if not isinstance(payload, dict):
        return {"value": payload, key: value}
    return {**payload, key: value}


def _delete_path(target: Any, segments: list[str]) -> None:
    if not isinstance(target, dict) or not segments:
        return
    head, *rest = segments
    if not rest:
        target.pop(head, None)
        return
    nxt = target.get(head)
    if isinstance(nxt, list):
        for item in nxt:
            _delete_path(item, rest)
        return
    _delete_path(nxt, rest)

File: packages/sdk/src/obligations.ts
/**
 * Egress obligation pipeline.
 *
 * Applies the obligations attached to a PDP decision (plus the static
 * Connection-Token obligations the runtime already merged in) to a reply
 * payload before it leaves the agent. The runtime does NOT currently apply
 * these obligations to outbound replies — that's the SDK's job on the
 * developer-facing `egress()` integration point.
 *
 * Supported obligation types (v0):
 *   - `redact_fields`        strip named keys (dot-path) from the payload
 *   - `redact_fields_except` keep only named keys (complement)
 *   - `redact_regex`         strip string values matching a regex
 *   - `summarize_only`       replace the payload with a `{summary}` marker
 *   - `aggregate_only`       replace rows with `{count, aggregate: 'count'}`
 *   - `insert_watermark`     attach a `_watermark` field
 *   - `no_downstream_share`  attach a `_no_downstream_share` marker
 *
 * Unknown obligation types are logged (via the caller's logger, if any) and
 * passed through — never silently dropped. Downstream adapters (or the
 * application) can layer on handling for the richer types (`rate_limit`,
 * `charge_usd`, `notify_principal`) which are not response-transforms.
 */

import type { Obligation } from '@kybernesis/arp-spec';

export interface ObligationApplyOptions {
  /** Called once per obligation type the runtime did not understand. */
  onUnknown?: (obligationType: string, params: Record<string, unknown>) => void;
}

/**
 * Apply the supplied obligations to `payload`. Returns a new object — never
 * mutates the input.
 */
export function applyObligations(
  payload: unknown,
  obligations: Obligation[],
  opts: ObligationApplyOptions = {},
): unknown {
  let current = cloneDeep(payload);
  for (const o of obligations) {
    current = applyOne(current, o, opts);
  }
  return current;
}

function applyOne(
  payload: unknown,
  obligation: Obligation,
  opts: ObligationApplyOptions,
): unknown {
  switch (obligation.type) {
    case 'redact_fields':
      return redactFields(payload, stringArray(obligation.params.fields));
    case 'redact_fields_except':
      return redactFieldsExcept(
        payload,
        stringArray(obligation.params.fields),
      );
    case 'redact_regex': {
      const pattern = typeof obligation.params.pattern === 'string'
        ? obligation.params.pattern
        : null;
      if (!pattern) return payload;
      return redactRegex(payload, pattern, typeof obligation.params.replacement === 'string'
        ? obligation.params.replacement
        : '[redacted]');
    }
    case 'summarize_only': {
      const max = typeof obligation.params.max_words === 'number'
        ? obligation.params.max_words
        : 50;
      return summarizeOnly(payload, max);
    }
    case 'aggregate_only':
      return aggregateOnly(payload);
    case 'insert_watermark':
      return insertWatermark(payload, obligation.params);
    case 'no_downstream_share':
      return addMarker(payload, '_no_downstream_share', true);
    // Non-response obligations (rate_limit, require_fresh_consent, etc.)
    // are enforced elsewhere in the pipeline; they pass through here.
    case 'rate_limit':
    case 'require_fresh_consent':
    case 'require_vc':
    case 'log_audit_level':
    case 'delete_after':
    case 'notify_principal':
    case 'charge_usd':
      return payload;
    default:
      opts.onUnknown?.(obligation.type, obligation.params);
      return payload;
  }
}

/* --------------------------- helpers --------------------------- */

function stringArray(value: unknown): string[] {
  if (!Array.isArray(value)) return [];
  return value.filter((v): v is string => typeof v === 'string');
}

export function redactFields(payload: unknown, fields: string[]): unknown {
  if (fields.length === 0) return payload;
  const out = cloneDeep(payload);
  for (const path of fields) {
    deletePath(out, path.split('.'));
  }
  return out;
}

export function redactFieldsExcept(
  payload: unknown,
  fields: string[],
): unknown {
  if (!isPlainObject(payload)) return payload;
  if (fields.length === 0) return {};
  const allow = new Set(fields);
  const keep: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(payload)) {
    if (allow.has(k)) keep[k] = v;
  }
  return keep;
}

export function redactRegex(
  payload: unknown,
  pattern: string,
  replacement: string,
): unknown {
  let re: RegExp;
  try {
    re = new RegExp(pattern, 'g');
  } catch {
    return payload;
  }
  return walkStrings(payload, (s) => s.replace(re, replacement));
}

function summarizeOnly(payload: unknown, maxWords: number): unknown {
  const text = typeof payload === 'string' ? payload : JSON.stringify(payload);
  const words = text.split(/\s+/).filter(Boolean);
  const summary = words.slice(0, maxWords).join(' ');
  return { summary, _obligation: 'summarize_only', max_words: maxWords };
}

function aggregateOnly(payload: unknown): unknown {
  if (Array.isArray(payload)) {
    return { count: payload.length, aggregate: 'count', _obligation: 'aggregate_only' };
  }
  if (isPlainObject(payload)) {
    for (const [k, v] of Object.entries(payload)) {
      if (Array.isArray(v)) {
        return { [k]: { count: v.length, aggregate: 'count' }, _obligation: 'aggregate_only' };
      }
    }
  }
  return { aggregate: 'count', count: 1, _obligation: 'aggregate_only' };
}

function insertWatermark(
  payload: unknown,
  params: Record<string, unknown>,
): unknown {
  return addMarker(payload, '_watermark', params);
}

function addMarker(payload: unknown, key: string, value: unknown): unknown {
  if (!isPlainObject(payload)) {
    return { value: payload, [key]: value };
  }
  return { ...payload, [key]: value };
}

function deletePath(target: unknown, segments: string[]): void {
  if (!isPlainObject(target) || segments.length === 0) return;
  const [head, ...rest] = segments;
  if (head === undefined) return;
  if (rest.length === 0) {
    delete (target as Record<string, unknown>)[head];
    return;
  }
  const next = (target as Record<string, unknown>)[head];
  if (Array.isArray(next)) {
    for (const item of next) deletePath(item, rest);
    return;
  }
  deletePath(next, rest);
}

function walkStrings(value: unknown, fn: (s: string) => string): unknown {
  if (typeof value === 'string') return fn(value);
  if (Array.isArray(value)) return value.map((v) => walkStrings(v, fn));
  if (isPlainObject(value)) {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(value)) out[k] = walkStrings(v, fn);
    return out;
  }
  return value;
}

function cloneDeep<T>(value: T): T {
  if (value === null || typeof value !== 'object') return value;
  if (Array.isArray(value)) return value.map(cloneDeep) as unknown as T;
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
    out[k] = cloneDeep(v);
  }
  return out as T;
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}
